packages/terra-overlay/src/Overlay.jsx

```diff
@@ -0,0 +1,168 @@
+  /**
+   * Center the content horizontally and vertically
+   */
+  centerContent: PropTypes.bool,
+  /**
+   * Visible overlay content when overlay is open
+   */
+  children: PropTypes.node,
+};
+
+const defaultProps = {
+  isOpen: false,
+  isFullScreen: true,
+  container: document.body,
+  centerContent: true,
+  canClickOutsideClose: true,
+};
+
+class Overlay extends React.Component {
+  constructor() {
+    super();
+    this.state = {};
+    this.trapFocus = this.trapFocus.bind(this);
+    this.handleEscKeyup = this.handleEscKeyup.bind(this);
+    this.handleClickOutside = this.handleClickOutside.bind(this);
+  }
+
+  componentDidMount() {
+    if (this.props.isOpen) {
+      this.onOpen();
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    if (!prevProps.isOpen && this.props.isOpen) {
+      this.onOpen();
+    } else if (prevProps.isOpen && !this.props.isOpen) {
+      this.onClose();
+    }
+  }
+
+  onOpen() {
+    overlayManager.add(this, this.props.container);
+    this.focus();
+    document.addEventListener('focus', this.trapFocus, true);
+    document.addEventListener('keyup', this.handleEscKeyup);
+  }
+
+  onClose() {
+    overlayManager.remove(this, this.props.container);
+    document.removeEventListener('focus', this.trapFocus, true);
+    document.removeEventListener('keyup', this.handleEscKeyup);
+    this.restoreFocus();
+  }
+
+  focus() {
+    this.focusBeforeOpen = document.activeElement;
+    this.overlay.firstElementChild.focus();
+  }
+
+  trapFocus() {
+    const currentActive = document.activeElement;
+    const child = this.overlay.firstElementChild;
+    if (child !== currentActive && !child.contains(currentActive)) {
+      child.focus();
+    }
+  }
+
+  restoreFocus() {
+    if (this.focusBeforeOpen && this.focusBeforeOpen.focus) {
+      this.focusBeforeOpen.focus();
+      this.focusBeforeOpen = null;
+    }
+  }
+
+  handleEscKeyup(e) {
+    if (e.keyCode === ESCAPE) {
+      this.props.onRequestClose();
+    }
+  }
+
+  handleClickOutside(e) {
+    if (this.props.canClickOutsideClose && e.target === e.currentTarget) {
+      this.props.onRequestClose();
+    }
+  }
+
+  render() {
+    const {
+      isOpen,
+      isFullScreen,
+      container,
+      onRequestClose,
+      canClickOutsideClose,
+      centerContent,
+      children,
+      ...customProps
+    } = this.props;
+    const overlayClass = classNames(
+      'terra-Overlay',
+      { 'terra-Overlay--center': centerContent },
+      customProps.className,
+    );
+    const onELementPosition = isFullScreen ? null : {
+      position: 'absolute',
+      top: container.scrollTop,
+      left: container.scrollLeft,
+    };
+    const overlayStyle = {
+      ...onELementPosition,
+      ...customProps.style,
+    };
+    /* eslint-disable jsx-a11y/no-static-element-interactions */
+    return (
+      <Portal isOpen={isOpen} target={container}>
+        <div
+          ref={(overlay) => { this.overlay = overlay; }}
```

Comment:
I think custom props should be spread onto this to remain consistent with the other components.

This does introduce a weird situation with ref and onClick. These will cancel out any custom ref and onClick as long as they're added after spreading customProps. Normally these functions would attempt to call the custom prop if one existed after performing the required actions on the component. So ref would call the custom prop ref as a callback, and onClick would call props.onClick as a callback (if they existed). 

This might be a situation where this component owns those actions specifically and doesn't callback, but I still think custom props should be used here. The callbacks probably should happen, but they can also always be opened up in the future.